packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+
+// Context:
+const PopupMenuContext = createContext<PopupMenuContextType | null>(null)
+
+// Hook to use the context:
+export const usePopupMenuContext = () => {
+  const context = useContext(PopupMenuContext)
+  if (!context) {
+    throw new Error(
+      "The usePopupMenuContext hook must be used inside within a PopupMenu that provides the PopupMenuContext."
+    )
+  }
+  return context
+}
+
+// Export the context for advanced use cases:
+export { PopupMenuContext }
+
+// ----- Component Definitions -----
+
+// POPUP MENU
+
+/** A Popup Menu component that wraps Headless UI Menu. The Menu will be rendered into a Juno Portal, so using Juno's PortalProvider (which is already included when using Juno's AppShell) is mandatory. */
+const PopupMenu: React.FC<PopupMenuProps> & {
+  Toggle: React.FC<PopupMenuToggleProps>
+  Menu: typeof PopupMenuMenu
+  Item: React.FC<PopupMenuItemProps>
+  Section: React.FC<PopupMenuSectionProps>
+  SectionHeading: React.FC<PopupMenuSectionHeadingProps>
+  SectionSeparator: React.FC<PopupMenuSectionSeparatorProps>
+} = ({
+  children = null,
+  className = "",
+  disabled = false,
+  icon = "moreVert",
+  menuSize = "normal",
+  onClose = undefined,
+  onOpen = undefined,
+  ...props
+}) => {
+  // Create a state to track headless-ui's internal open state (Note that we can track the internal state and can let headless ui update our tracking state, but we can not set it, hence there is no `open` prop on our component):
+  const [isOpen, setIsOpen] = useState(false)
+  const previousOpenRef = useRef(isOpen) // Track the previous value of isOpen for efficiency
+
+  // Run handlers when our tracking state changes based on changes of the internal headless state, or when the handlers themselves change:
+  useEffect(() => {
+    if (isOpen !== previousOpenRef.current) {
+      if (isOpen) {
+        onOpen?.()
+      } else {
+        onClose?.()
+      }
+    }
+    previousOpenRef.current = isOpen // Update the previous state value
+  }, [isOpen, onOpen, onClose])
+
+  // Ensure we always have an array of children, even if no or only a single child were passed:
+  const childrenArray = React.Children.toArray(children)
+
+  // Test whether a PopupMenu.Toggle child was passed:
+  const hasToggle = childrenArray.some((child) => React.isValidElement(child) && child.type === PopupMenuToggle)
+
+  // Test whether a PopupMenu.Menu child was passed:
+  const menu = childrenArray.find((child) => React.isValidElement(child) && child.type === PopupMenuMenu)
+
+  return (
+    <HeadlessMenu as="div" className={`juno-popupmenu ${className}`} {...props}>
+      {/* Update our open state when the open render prop from headless ui menu changes, so we can run the handlers when our tracking state changes: */}
+      {({ open, close }) => {
+        // Set our tracking open state outside of rendering cycle. Trying to set the state during render will cause an error:
+        useEffect(() => {
```

Comment:
This is a big part in our next refactoring sessions, we need to get rid of this unnecessary use effects, I believe that you could just avoid it here and use simple setState here.

Reply:
I believe we mentioned this in the "improvements after migration" epic, right? Wouldn't mess with what we know works in the context of this PR

Reply:
Generally I do agree with using fewer `useEffect`s, we may have gone a bit overboard with these in the past, however this case is special:

In order to update our tracking state, we need to rely on the render props as provided by Headless menu. There is no other way of knowing whether the menu is open or not. This render prop updates in the components render cycle, using a useEffect as in an anonymous function that recieves the render props from headless menu in an object as an argument returned by the component, and again is used to update our tracking state. This is a valid (and the only afaik) way to achieve this, so this useEffect is crucial for the functionality of the component. (Also see next comment).